cheetah/accelerator/space_charge_kick.py

```diff
@@ -0,0 +1,684 @@
+        p = torch.sqrt(
+            moments[..., 1] ** 2 + moments[..., 3] ** 2 + moments[..., 5] ** 2
+        )
+        gamma = torch.sqrt(1 + (p / (electron_mass * speed_of_light)) ** 2)
+
+        particles[..., 1] = moments[..., 1] / p0.unsqueeze(-1)
+        particles[..., 3] = moments[..., 3] / p0.unsqueeze(-1)
+        particles[..., 4] = -moments[..., 4] / betaref.unsqueeze(-1)
+        particles[..., 5] = (gamma - gammaref.unsqueeze(-1)) / (
+            (betaref * gammaref).unsqueeze(-1)
+        )
+
+        return particles
+
+    def _compute_forces(
+        self,
+        beam: ParticleBeam,
+        moments: torch.Tensor,
+        cell_size: torch.Tensor,
+        grid_dimensions: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Interpolates the space charge force from the grid onto the macroparticles.
+        Reciprocal function of _deposit_charge_on_grid.
+        """
+        grad_x, grad_y, grad_z = self._E_plus_vB_field(
+            beam, moments, cell_size, grid_dimensions
+        )
+        grid_shape = self.grid_shape
+        interpolated_forces = torch.zeros(
+            (*beam.particles.shape[:-1], 3), **self.factory_kwargs
+        )
+
+        # Get particle positions
+        particle_positions = moments[..., [0, 2, 4]]
+        normalized_positions = (
+            particle_positions + grid_dimensions.unsqueeze(-2)
+        ) / cell_size.unsqueeze(-2)
+
+        # Find indices of the lower corners of the cells containing the particles
+        cell_indices = torch.floor(normalized_positions).type(torch.long)
+
+        # Calculate the weights for all surrounding cells
+        offsets = torch.tensor(
+            [
+                [0, 0, 0],
+                [0, 0, 1],
+                [0, 1, 0],
+                [0, 1, 1],
+                [1, 0, 0],
+                [1, 0, 1],
+                [1, 1, 0],
+                [1, 1, 1],
+            ]
+        )
+        surrounding_indices = cell_indices.unsqueeze(-2) + offsets.unsqueeze(
+            -3
+        )  # Shape:(.., num_particles, 8, 3)
+        weights = 1 - torch.abs(
+            normalized_positions.unsqueeze(-2) - surrounding_indices
+        )  # Shape: (..., num_particles, 8, 3)
+        cell_weights = weights.prod(dim=-1)  # Shape: (..., num_particles, 8)
+
+        # Extract forces from the grids
+        surrounding_indices_flattened = surrounding_indices.flatten(
+            start_dim=-3, end_dim=-2
+        )  # Shape: (..., num_particles * 8, 3)
+        idx_x = surrounding_indices_flattened[..., 0]
+        idx_y = surrounding_indices_flattened[..., 1]
+        idx_s = surrounding_indices_flattened[..., 2]
+        valid_mask = (
+            (idx_x >= 0)
+            & (idx_x < grid_shape[0])
+            & (idx_y >= 0)
+            & (idx_y < grid_shape[1])
+            & (idx_s >= 0)
+            & (idx_s < grid_shape[2])
+        )
+
+        # TODO: Is there a better alternative to this loop? (maybe torch.vmap?)
```

Comment:
This loop is a bit of a hack to achieve functionality like vectorisation, but of course it's not true vectorisation.

More importantly, it breaks the gradient computations. Ideally, we would find a way to properly vectorise it like the rest. I couldn't immediately come up with something, but that might also due to my lack of understanding exactly what is going on. In the worst case, we could try to fix it with `torch.vmap`.